src/restic/node.go

```diff
@@ -331,11 +321,26 @@ func (node *Node) UnmarshalJSON(data []byte) error {
 	type nodeJSON Node
 	nj := (*nodeJSON)(node)
 
-	err := json.Unmarshal(data, nj)
-	if err != nil {
-		return errors.Wrap(err, "Unmarshal")
+	if err := json.Unmarshal(data, nj); err != nil {
+		// Retry conversion with list of ExtendedAttributes
```

Comment:
For this special case I'd suggest a unit test for just the `Unmarshal()` function.

> That's done in integration testing, right? if it can still read small-repo.tar.gz and consorts, the legacy path is fine.

That depends on the code, I'm afraid. There are several archived repos that are tested, but there isn't any explicit test if the extended attributes of a node are restored correctly. So the failure case "extended attributes in the old format are not read at all, it always returns an empty map" won't be detected.

> Also, from a bigger picture, isn't this the best way forward? How else would you make json effecting changes?

Very, very carefully. If in doubt I think we should postpone optimizations for a future repository version. For the repo v2 work, I'll add an `upgrade` command (name tbd) that will migrate the data to the new format.

For all changes of the JSON format I'd like to have solid reasoning, there's a lot that may go wrong. Adding things is usually easy (e.g. introducing extended attributes at all), but changing things offers a lot of room for bugs and errors.